### ISSUE TITLE
关于3.1 Centrifugal Reference Frame 代码

### ISSUE THREAD

**Issue by Cauchyck** [yokinglou/CRIN#2]
您好，不知道能不能先上传3.1节的代码以及图3的可视化代码？

**Comment by Cauchyck** [comment]
如果不方便的话能否帮我看下我根据论文复现的代码哪里有错？
def Q_normal(Q):
    # 计算邻域矩阵 W
    diffs = Q.unsqueeze(1) - Q.unsqueeze(0)
    dists = torch.norm(diffs, dim=-1)
    sigma = torch.median(dists)
    weights = torch.exp(-dists ** 2 / sigma ** 2)
    WN = normalize(weights, p=1, dim=1)

    # 计算法向量
    diffs = diffs.permute(1, 0, 2)
    WN = WN.unsqueeze(-1)
    cov = torch.matmul(diffs.permute(0, 2, 1), diffs * WN)
    eigvals, eigvecs = torch.symeig(cov, eigenvectors=True)
    normals = eigvecs[:, :, 0]

    mask = normals[:, 2] < 0
    mask = mask.unsqueeze(-1)
    normals = torch.where(mask, -normals, normals)

    return normals


def calc_B(Q, normal=False):
    # Q is a tensor of shape [M, 3], M is the number of points
    if normal:
        Q = Q_normal(Q)
    # w is a tensor of shape [M, 3], each row is q / norm(q)

    w = Q / torch.norm(Q, dim=1, keepdim=True)

    # z is a tensor of shape [M, 3], each row is [0, 0, 1]
    z = torch.tensor([[0.0, 0.0, 1.0]], device=Q.device) * Q[:, 2:3]

    # u is a tensor of shape [M, 3], each row is z * w / norm(z * w)
    u = torch.cross(z, w) / torch.norm(torch.cross(z, w), dim=1, keepdim=True)

    # v is a tensor of shape [M, 3], each row is w * u
    v = torch.cross(w, u)

    # B_q is a tensor of shape [M, 3, 3], each slice along dim=0 is [u,v,w]
    B = torch.stack([u, v, w], dim=1).transpose(1, 2)

    return B

**Comment by yokinglou** [comment]
可以在这里查看CRF代码[CRF dev code](https://github.com/yokinglou/CRIN/blob/dev/models/modules.py#L5)

**Comment by Cauchyck** [comment]
好的，谢谢！能否再给一下计算normal的代码。另外，源点云的crf和对源点云进行随机旋转后再求的crf在数值上是否一样？
🙏🙏🙏

**Comment by yokinglou** [comment]
法向量估计可以使用Open3D提供的API，或者直接使用ModelNet和ShapeNet数据集提供的法向量。

旋转前后同一个点的CRF是不同的，因为该点的坐标也发生了变化，这可以从定义发现。不变的是点云转换到CRF后点的坐标。

**Comment by Cauchyck** [comment]
好的。我的问题是这样的，对于点p，可以求得一个crf_p。设点q为旋转R后的p，那么在数值上，crf_p(p) = crp_(q)是否成立

**Comment by yokinglou** [comment]
r(Rp)=r(p)是指点p经过旋转前后，在各自CRF当中的坐标相等。可以这么理解，经过旋转后，不仅点p的坐标发生了变化，其对应的CRF的基底B也是发生了变化，但是两者相乘是保持不变的。可以参考论文附录当中的A.3。

**Comment by Cauchyck** [comment]
好的，我理解了，感谢您的讲解
